File: 2016/2016_5.py
from hashlib import md5

def solve_part_1(door_id):
    idx = 0

    password = []

    while len(password) < 8:
        if idx % 1000000 == 0: print(idx)

        m = md5()
        m.update(door_id)
        m.update(bytearray(str(idx), 'utf-8'))
        digest = m.hexdigest()

        if digest[:5] == '00000':
            print('***' + digest[5])
            password.append(digest[5])

        idx += 1

    return ''.join(password)

def solve_part_2(door_id):
    idx = 0

    password = [None] * 8

    while len(list(filter(lambda x: x is None, password))) > 0:
        if idx % 1000000 == 0: print(idx)

        m = md5()
        m.update(door_id)
        m.update(bytearray(str(idx), 'utf-8'))
        digest = m.hexdigest()

        if digest[:5] == '00000':
            try:
                password_idx = int(digest[5])
            except ValueError:
                idx += 1
                continue

            if password_idx > 7 or password[password_idx] is not None:
                idx += 1
                continue
                
            print('index ' + digest[5])
            print('letter ' + digest[6])
            password[password_idx] = digest[6]

        idx += 1

    return ''.join(password)

#print(solve_part_1('abbhdwsy'))
print(solve_part_2('abbhdwsy'))

File: 2016/2016_5.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <CommonCrypto/CommonDigest.h>

char *str2md5(const char *str, int length) {
    int n;
    CC_MD5_CTX c;
    unsigned char digest[16];
    char *out = (char*)malloc(33);

    CC_MD5_Init(&c);

    while (length > 0) {
        if (length > 512) {
            CC_MD5_Update(&c, str, 512);
        } else {
            CC_MD5_Update(&c, str, length);
        }
        length -= 512;
        str += 512;
    }

    CC_MD5_Final(digest, &c);

    for (n = 0; n < 16; ++n) {
        snprintf(&(out[n*2]), 16*2, "%02x", (unsigned int)digest[n]);
    }

    return out;
}

char* solve_part_1(const char* door_id) {
	int idx = 0;

	char *password = calloc(9, sizeof(char));

	int password_idx = 0;

	while (password_idx < 8) {
		if (idx % 1000000 == 0) {
			printf("%d\n", idx);
		}

		char door_id_with_idx[20];

		sprintf(door_id_with_idx, "%s%d", door_id, idx);

		char* door_id_md5 = str2md5(door_id_with_idx, strlen(door_id_with_idx));

		if (door_id_md5[0] == '0' && door_id_md5[1] == '0' &&\
		 door_id_md5[2] == '0' && door_id_md5[3] == '0' &&\
		 door_id_md5[4] == '0') {
			printf("*** %c\n", door_id_md5[5]);
			password[password_idx] = door_id_md5[5];
			password_idx++;
		}

		idx++;
	}

	return password;
}

int main(int argc, char* argv[]) {
	printf("%s\n", solve_part_1("abbhdwsy"));
}
